litex/soc/cores/i2s.py

```diff
@@ -413,53 +538,85 @@ def __init__(self, pads, fifo_depth=256):
                 )
             ]
 
-            tx_cnt = Signal(5)
-            tx_buf = Signal(32)
+            tx_buf_width = fifo_data_width + 1 if frame_format == I2S_FORMAT.I2S_STANDARD else fifo_data_width
+            sample_width = sample_width + 1 if frame_format == I2S_FORMAT.I2S_STANDARD else sample_width
+            offset = [0] if frame_format == I2S_FORMAT.I2S_STANDARD else []
+
+            tx_cnt_width = math.ceil(math.log(fifo_data_width,2))
+            tx_cnt = Signal(tx_cnt_width)
+            tx_buf = Signal(tx_buf_width)
             self.submodules.txi2s = txi2s = FSM(reset_state="IDLE")
             txi2s.act("IDLE",
                 If(self.tx_ctl.fields.enable,
-                    If(falling_edge & sync_pin,
+                    If(falling_edge & (~sync_pin if frame_format == I2S_FORMAT.I2S_STANDARD else sync_pin),
                         NextState("WAIT_SYNC"),
                     )
                 )
             ),
             txi2s.act("WAIT_SYNC",
-                If(falling_edge & ~sync_pin,
+                If(falling_edge & (~sync_pin if frame_format == I2S_FORMAT.I2S_STANDARD else sync_pin),
                     NextState("LEFT"),
-                    NextValue(tx_cnt, 16),
-                    NextValue(tx_buf, tx_rd_d),
-                    tx_rden.eq(1)
+                    NextValue(tx_cnt, sample_width),
+                    NextValue(tx_buf, Cat(tx_rd_d, offset)),
+                    tx_rden.eq(1),
                 )
             )
             txi2s.act("LEFT",
                 If(~self.tx_ctl.fields.enable,
                     NextState("IDLE")
                 ).Else(
-                    NextValue(tx_pin, tx_buf[31]),
+                    NextValue(tx_pin, tx_buf[sample_width - 1]),
```

Comment:
if concatenate_channels is true, in this case, you're reading from tx_buf[15] when sample_width is 16. This causes left data to go to the right channel and then the left channel gets silence. 

Suggest adding a new variable `sample_msb` that is set to `sample_width * 2` and using it when the indexing into tx_buf.

Reply:
As suggested we introduced `sample_msb` variable which is equal to `fifo_data_width - 1`.

For concatenated channels `fifo_data_width` is equal to `sample_width * 2`, so it should work as expected.